crates/pathfinder/examples/feeder_gateway.rs

```diff
@@ -0,0 +1,399 @@
+use std::collections::HashMap;
+
+use anyhow::Context;
+use pathfinder_common::{
+    Chain, ClassHash, ContractAddress, ContractNonce, StarknetBlockHash, StarknetBlockNumber,
+};
+use pathfinder_ethereum::contract::ContractAddresses;
+use pathfinder_storage::{
+    StarknetBlocksBlockId, StarknetBlocksTable, StarknetStateUpdatesTable,
+    StarknetTransactionsTable,
+};
+use serde::Deserialize;
+use stark_hash::Felt;
+use starknet_gateway_types::reply::{
+    state_update::{DeclaredSierraClass, DeployedContract, ReplacedClass, StorageDiff},
+    Status,
+};
+use tracing_subscriber::{fmt, prelude::*, EnvFilter};
+use warp::Filter;
+
+/// Serve feeder gateway REST endpoints required for pathfinder to sync.
+///
+/// Usage:
+/// `cargo run --release -p pathfinder --example feeder_gateway ./testnet2.sqlite`
+///
+/// Then pathfinder can be run with the following arguments to use this tool as a sync source:
+///
+/// ```text
+/// cargo run --release -p pathfinder -- \
+///     --network custom --chain-id SN_GOERLI2 \
+///     --gateway-url http://localhost:8080/gateway \
+///     --feeder-gateway-url http://localhost:8080/feeder_gateway \
+///     --data-directory /tmp
+/// ```
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    tracing_subscriber::registry()
+        .with(fmt::layer())
+        .with(EnvFilter::from_default_env())
+        .init();
+    serve().await
+}
+
+async fn serve() -> anyhow::Result<()> {
+    let database_path = std::env::args().nth(1).unwrap();
+    let storage = pathfinder_storage::Storage::migrate(
+        database_path.into(),
+        pathfinder_storage::JournalMode::WAL,
+    )?;
+
+    let chain = {
+        let mut connection = storage.connection()?;
+        let tx = connection.transaction()?;
+        get_chain(&tx)?
+    };
+
+    let get_contract_addresses = warp::path("get_contract_addresses").map(move || {
+        let addresses = contract_addresses(chain).unwrap();
+        let reply =
+            serde_json::json!({"GpsStatementVerifier": addresses.gps, "Starknet": addresses.core});
+        warp::reply::json(&reply)
+    });
+
+    #[derive(Debug, Deserialize)]
+    struct BlockIdParam {
+        #[serde(default, rename = "blockNumber")]
+        block_number: Option<String>,
+        #[serde(default, rename = "blockHash")]
+        block_hash: Option<StarknetBlockHash>,
+    }
+
+    impl TryInto<StarknetBlocksBlockId> for BlockIdParam {
+        type Error = ();
+
+        fn try_into(self) -> Result<StarknetBlocksBlockId, Self::Error> {
+            if let Some(n) = self.block_number {
+                if n == "latest" {
+                    return Ok(StarknetBlocksBlockId::Latest);
+                } else {
+                    let n: u64 = n.parse().map_err(|_| ())?;
+                    return Ok(StarknetBlocksBlockId::Number(
+                        StarknetBlockNumber::new_or_panic(n),
+                    ));
+                }
```

Comment:
What about `pending`? 😁 I realise we can't support that, but it should maybe get handled by just not returning anything of value, instead of error'ing? Although not too important.